readthedocs/vcs_support/backends/git.py

```diff
@@ -199,6 +200,35 @@ def clone(self):
             raise RepositoryError
         return code, stdout, stderr
 
+    @property
+    def lsremote(self):
+        """
+        Use ``git ls-remote`` to list branches and tags without clonning the repository.
+
+        :returns: tuple containing a list of branch and tags
+        """
+        cmd = ['git', 'ls-remote', self.repo_url]
+
+        code, stdout, stderr = self.run(*cmd)
+        if code != 0:
+            raise RepositoryError
+
+        tags = []
+        branches = []
+        for line in stdout.splitlines()[1:]:  # skip HEAD
+            commit, ref = line.split()
+            if ref.startswith('refs/heads/'):
+                branch = ref.replace('refs/heads/', '')
+                branches.append(VCSVersion(self, branch, branch))
+            if ref.startswith('refs/tags/'):
+                tag = ref.replace('refs/tags/', '')
+                if tag.endswith('^{}'):
+                    # skip annotated tags since they are duplicated
+                    continue
```

Comment:
This formatting is weird, but it is git's default.

```
$ git ls-remote https://github.com/readthedocs/test-builds | grep annotated
886108064d63dde44c14f657c994c63233e9c3ab	refs/tags/annotated-tag
d39922ff32b8bc5c0182e7dd976fd921dba0d873	refs/tags/annotated-tag^{}
```

`^` is a reserved character and branches/tags can't use it. See https://git-scm.com/docs/git-check-ref-format